### ISSUE TITLE
mockk throws ClassNotFoundException: io.mockk.proxy.jvm.dispatcher.JvmMockKDispatcher when run with Spek2

### ISSUE THREAD

**Issue by snowe2010** [mockk/mockk#239]
Internal error is thrown when attempting to mock class in Spek2: 

When running `SmallObjectTest` with this code.
https://github.com/snowe2010/pretty-print/tree/0b3831edd67cab1b894df01df2cfe49d42e20cae

#### Steps to Reproduce

1. run `SmallObjectTest` with this commit https://github.com/snowe2010/pretty-print/tree/0b3831edd67cab1b894df01df2cfe49d42e20cae
(might need to run in intellij with latest kotlin and spek plugins?)

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

All versions are in the provided gradle.build.kts, but I'll repeat them here. 

* MockK version: 1.9.kotlin12
* OS: Windows 10 Pro : 10.0.17763 Build 17763
* Kotlin version: 1.3.11
* JDK version: 1.8
* JUnit version: (the one provided with kotlin? `testCompile(kotlin("test-junit"))`
* Type of test: unit test OR android instrumented test: Spek2 unit test

#### Failure Logs

```
Testing started at 9:31 AM ...
"C:\Program Files\Java\jdk1.8.0_121\bin\java.exe" -javaagent:C:\Users\Tyler\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\183.5429.1\lib\idea_rt.jar=2598:C:\Users\Tyler\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\183.5429.1\bin -Dfile.encoding=UTF-8 -classpath "C:\Program Files\Java\jdk1.8.0_121\jre\lib\charsets.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\deploy.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\access-bridge-64.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\cldrdata.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\dnsns.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\jaccess.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\jfxrt.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\localedata.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\nashorn.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\sunec.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\sunjce_provider.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\sunmscapi.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\sunpkcs11.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\ext\zipfs.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\javaws.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\jce.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\jfr.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\jfxswt.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\jsse.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\management-agent.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\plugin.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\resources.jar;C:\Program Files\Java\jdk1.8.0_121\jre\lib\rt.jar;C:\Users\Tyler\IdeaProjects\pretty-print\out\test\classes;C:\Users\Tyler\IdeaProjects\pretty-print\out\production\classes;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-jdk8\1.3.11\dd9bff00d6cfca58b6c1fe89be8e0678e35cf35f\kotlin-stdlib-jdk8-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk\1.9.kotlin12\31e2bd4eac2f0dd878b78d70bc88fe7b38c332f5\mockk-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk-dsl-jvm\1.9.kotlin12\4df110c2cb099d8522eae74de140ab77465c9b8f\mockk-dsl-jvm-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-reflect\1.3.11\aae7b33412715e9ed441934c4ffaad1bb80e9d36\kotlin-reflect-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-test-junit\1.3.11\955bf353ab1459d9661e6512305417f9474171d9\kotlin-test-junit-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-test\1.3.11\c68def89e3abacb012d020bd85dc685e2c7fd34d\kotlin-test-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.spekframework.spek2\spek-dsl-jvm\2.0.0-alpha.2\4c7cc80c80cb92e2daa3d4b3f393e9ebc214db84\spek-dsl-jvm-2.0.0-alpha.2.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-jdk7\1.3.11\4839661cf6ce3c14b65ed7dcf5b9249b44ecca16\kotlin-stdlib-jdk7-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib\1.3.11\4cbc5922a54376018307a731162ccaf3ef851a39\kotlin-stdlib-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-test-common\1.3.11\42db8fc7cd17ec8278d06862d8cf3b4028675e84\kotlin-test-common-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.spekframework.spek2\spek-runner-junit5\2.0.0-alpha.2\131561375b57ba786f8a7de6827803baa23cf13c\spek-runner-junit5-2.0.0-alpha.2.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.spekframework.spek2\spek-runtime-jvm\2.0.0-alpha.2\eded81270e783ec6fee1a7bf87fec5a592310711\spek-runtime-jvm-2.0.0-alpha.2.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-test-annotations-common\1.3.11\3dcddf0db7082ff348c31f44499d075173a5498f\kotlin-test-annotations-common-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\junit\junit\4.12\2973d150c0dc1fefe998f834810d68f278ea58ec\junit-4.12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk-common\1.9.kotlin12\fb622dadcd0bc398f067ef935c65f5e7f29d0ff\mockk-common-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk-agent-jvm\1.9.kotlin12\3f39fdb866e96325165f01ec3f3a66e1b970611c\mockk-agent-jvm-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.spekframework.spek2\spek-dsl-common\2.0.0-alpha.2\d512538fdcc6449cf09899806be56c7b3717f84\spek-dsl-common-2.0.0-alpha.2.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains.kotlin\kotlin-stdlib-common\1.3.11\d8b8e746e279f1c4f5e08bc14a96b82e6bb1de02\kotlin-stdlib-common-1.3.11.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.jetbrains\annotations\13.0\919f0dfe192fb4e063e7dacadee7f8bb9a2672a9\annotations-13.0.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.hamcrest\hamcrest-core\1.3\42a25dc3219429f0e5d060061f71acb49bf010a0\hamcrest-core-1.3.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk-dsl\1.9.kotlin12\da6de0efb9c9fdc315917fa10a0e8032b7846e75\mockk-dsl-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk-agent-common\1.9.kotlin12\7542834dab87aa24df2d8e3b29bbcdcdc1bc2aca\mockk-agent-common-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.spekframework.spek2\spek-runtime-common\2.0.0-alpha.2\66b95c2c2a76abf62b0354e7e290e7b05895a57\spek-runtime-common-2.0.0-alpha.2.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.mockk\mockk-agent-api\1.9.kotlin12\ac930075957aeb7364b8d568d3db5b5e1b7bd2ae\mockk-agent-api-1.9.kotlin12.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\org.objenesis\objenesis\2.6\639033469776fd37c08358c6b92a4761feb2af4b\objenesis-2.6.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\net.bytebuddy\byte-buddy\1.9.3\f32e510b239620852fc9a2387fac41fd053d6a4d\byte-buddy-1.9.3.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\net.bytebuddy\byte-buddy-agent\1.9.3\f5b78c16cf4060664d80b6ca32d80dca4bd3d264\byte-buddy-agent-1.9.3.jar;C:\Users\Tyler\.gradle\caches\modules-2\files-2.1\io.github.classgraph\classgraph\4.0.6\7fe1e7464ffd40c7333b8bbf53c4d73a7cbd7b4c\classgraph-4.0.6.jar;C:\Users\Tyler\.IntelliJIdea2018.3\config\plugins\Spek Framework\lib\interop-jvm-2.0.0-rc.1.180+b8533a4-all.jar" org.spekframework.ide.ConsoleKt --sourceDirs C:\Users\Tyler\IdeaProjects\pretty-print\out\test\classes --paths Y29tLnR5bGVydGhyYWlsa2lsbC5oZWxwZXJzLnByZXR0eXByaW50/U21hbGxPYmplY3RUZXN0

Exception in thread "main" java.lang.NoClassDefFoundError: io/mockk/proxy/jvm/dispatcher/JvmMockKDispatcher
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at io.mockk.proxy.jvm.transformation.SubclassInstrumentation$AdviceBuilder.build(SubclassInstrumentation.kt:28)
	at io.mockk.proxy.jvm.transformation.SubclassInstrumentation.<init>(SubclassInstrumentation.kt:33)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory$init$Initializer.init(JvmMockKAgentFactory.kt:92)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.init(JvmMockKAgentFactory.kt:123)
	at io.mockk.impl.JvmMockKGateway.<init>(JvmMockKGateway.kt:46)
	at io.mockk.impl.JvmMockKGateway.<clinit>(JvmMockKGateway.kt:169)
	at com.tylerthrailkill.helpers.prettyprint.SmallObjectTest$1.invoke(SmallObjectTest.kt:28)
	at com.tylerthrailkill.helpers.prettyprint.SmallObjectTest$1.invoke(SmallObjectTest.kt:9)
	at org.spekframework.spek2.runtime.AbstractRuntime.resolveSpec(SpekRuntime.kt:27)
	at org.spekframework.spek2.runtime.SpekRuntime.discover(SpekJvmRuntime.kt:35)
	at org.spekframework.ide.Spek2ConsoleLauncher.run(console.kt:21)
	at org.spekframework.ide.ConsoleKt$main$1.invoke(console.kt:34)
	at org.spekframework.ide.ConsoleKt$main$1.invoke(console.kt)
	at shadow.com.xenomachina.argparser.SystemExitExceptionKt.mainBody(SystemExitException.kt:74)
	at shadow.com.xenomachina.argparser.SystemExitExceptionKt.mainBody$default(SystemExitException.kt:72)
	at org.spekframework.ide.ConsoleKt.main(console.kt:32)
Caused by: java.lang.ClassNotFoundException: io.mockk.proxy.jvm.dispatcher.JvmMockKDispatcher
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 40 more

Process finished with exit code 1
```

**Comment by catchin** [comment]
I have a similar problem, but I cannot provide the sample right now.

**Context**

-   MockK version: 1.9
-   OS: Linux inside docker
-  Kotlin version: 1.3.11
-  JDK version: 1.8
-  JUnit version: 5
-  Type of test: unit test

**Stack trace:**

```my.package.SomeTest > initializationError FAILED
     java.lang.NoClassDefFoundError: io/mockk/proxy/jvm/dispatcher/JvmMockKDispatcher
         at java.lang.ClassLoader.defineClass1(Native Method)
         at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
         at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
         at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
         at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
         at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
         at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
         at java.security.AccessController.doPrivileged(Native Method)
         at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
         at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
         at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
         at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
         at java.lang.ClassLoader.defineClass1(Native Method)
         at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
         at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
         at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
         at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
         at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
         at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
         at java.security.AccessController.doPrivileged(Native Method)
         at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
         at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
         at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
         at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
         at io.mockk.proxy.jvm.transformation.SubclassInstrumentation$AdviceBuilder.build(SubclassInstrumentation.kt:28)
         at io.mockk.proxy.jvm.transformation.SubclassInstrumentation.<init>(SubclassInstrumentation.kt:33)
         at io.mockk.proxy.jvm.JvmMockKAgentFactory$init$Initializer.init(JvmMockKAgentFactory.kt:92)
         at io.mockk.proxy.jvm.JvmMockKAgentFactory.init(JvmMockKAgentFactory.kt:123)
         at io.mockk.impl.JvmMockKGateway.<init>(JvmMockKGateway.kt:46)
         at io.mockk.impl.JvmMockKGateway.<clinit>(JvmMockKGateway.kt:169)
         at my.package.SomeTest.<init>(SomeTest.kt:103)
 
         Caused by:
         java.lang.ClassNotFoundException: io.mockk.proxy.jvm.dispatcher.JvmMockKDispatcher
             at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
             at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
             at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
             at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
             ... 31 more```

**Comment by oleksiyp** [comment]
Hm, is it JDK or JVM? May be the case that JDK is needed. It is weird. It means ByteBuddy was not able to load properly I believe. Can you enable TRACE logging with logback? And send here

**Comment by snowe2010** [comment]
My snippet contained the classpath. It is jdk. 

With logback trace
```
TRACE - Starting JVM MockK implementation. Java version = 1.8.0_121. 
TRACE - Byte buddy agent installed
TRACE - Error creating boot jar
java.io.IOException: Access is denied
	at java.io.WinNTFileSystem.createFileExclusively(Native Method)
	at java.io.File.createTempFile(File.java:2024)
	at java.io.File.createTempFile(File.java:2070)
	at io.mockk.proxy.jvm.dispatcher.BootJarLoader.buildBootJar(BootJarLoader.kt:54)
	at io.mockk.proxy.jvm.dispatcher.BootJarLoader.loadBootJar(BootJarLoader.kt:18)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.initInstrumentation(JvmMockKAgentFactory.kt:139)
	at io.mockk.proxy.jvm.JvmMockKAgentFactory.init(JvmMockKAgentFactory.kt:34)
	at io.mockk.impl.JvmMockKGateway.<init>(JvmMockKGateway.kt:46)
	at io.mockk.impl.JvmMockKGateway.<clinit>(JvmMockKGateway.kt:169)
	at com.tylerthrailkill.helpers.prettyprint.SmallObjectTest$1.invoke(SmallObjectTest.kt:30)
	at com.tylerthrailkill.helpers.prettyprint.SmallObjectTest$1.invoke(SmallObjectTest.kt:11)
	at org.spekframework.spek2.runtime.AbstractRuntime.resolveSpec(SpekRuntime.kt:27)
	at org.spekframework.spek2.runtime.SpekRuntime.discover(SpekJvmRuntime.kt:35)
	at org.spekframework.ide.Spek2ConsoleLauncher.run(console.kt:21)
	at org.spekframework.ide.ConsoleKt$main$1.invoke(console.kt:34)
	at org.spekframework.ide.ConsoleKt$main$1.invoke(console.kt)
	at shadow.com.xenomachina.argparser.SystemExitExceptionKt.mainBody(SystemExitException.kt:74)
	at shadow.com.xenomachina.argparser.SystemExitExceptionKt.mainBody$default(SystemExitException.kt:72)
	at org.spekframework.ide.ConsoleKt.main(console.kt:32)
TRACE - Can't inject boot jar.
```

[Reproduction branch here](https://github.com/snowe2010/pretty-print/tree/mockk-issue-239)

**Comment by oleksiyp** [comment]
Looks like your temporary directory is not writeable. Just point Java temporary directory to something that it is possible to create file at.

**Comment by catchin** [comment]
Thanks @oleksiyp, looks like this caused the problem. Setting the temporary directory in gradle test task via `systemProperty "java.io.tmpdir", "some-temp-dir"` (and making sure it exists) fixed the problem for me.

**Comment by rooton** [comment]
The same issue for me

**Comment by oleksiyp** [comment]
In 5aa43c2 I added one more try to create a file in the current directory in case tmp is not available. Not sure if it will make life easier for you

**Comment by oleksiyp** [comment]
Released [1.9.1](https://github.com/mockk/mockk/releases/tag/1.9.1). Please try it out(as I said it may not help) and close the ticket if everything is fine

**Comment by fellahi-ali** [comment]
> Looks like your temporary directory is not writeable. Just point Java temporary directory to something that it is possible to create file at.

Exactly that is the root of the problem.
This issue is in the Spek plugin and discussed [here]( https://github.com/spekframework/spek/issues/563)
There is a workaround [here](https://github.com/spekframework/spek/issues/563#issuecomment-459313727) and the issue is resolved by Spek team and will be available in 2.0.1 release.

**Comment by fellahi-ali** [comment]
> Thanks @oleksiyp, looks like this caused the problem. Setting the temporary directory in gradle test task via `systemProperty "java.io.tmpdir", "some-temp-dir"` (and making sure it exists) fixed the problem for me.

Another solution can be found [here](https://github.com/spekframework/spek/issues/563#issuecomment-459313727)

**Comment by fellahi-ali** [comment]
> The same issue for me

Solution can be found [here](https://github.com/spekframework/spek/issues/563#issuecomment-459313727)

**Comment by fellahi-ali** [comment]
> Released [1.9.1](https://github.com/mockk/mockk/releases/tag/1.9.1). Please try it out(as I said it may not help) and close the ticket if everything is fine

Thanks, I had that issue, and this works for me.
